EIPS/eip-3540.md

```diff
@@ -0,0 +1,218 @@
+
+## Motivation
+
+On-chain deployed EVM bytecode contains no pre-defined structure today. Code is typically validated in clients to the extent of `JUMPDEST` analysis at runtime, every single time prior to execution. This poses not only an overhead, but also a challenge for introducing new or deprecating old features. [This initial proposal](https://notes.ethereum.org/@axic/evm-object-format) explains some of these challenges.
+
+Validating code during the contract creation process allows code versioning without an additional version field in the account. Versioning is a useful tool for introducing or deprecating features, especially for larger changes (such as significant changes to control flow, or features like account abstraction).
+
+The format described in this EIP introduces a simple and extensible container with a minimal set of changes required to both clients and languages, and introduces validation.
+
+The first tangible feature it provides is separation of code and data. This separation is especially beneficial for on-chain code validators (like those utilised by layer-2 scaling tools, such as Optimism), because they can more efficiently validate contracts (for example by skipping the data section). This can result in significant gas savings. Additionally, various (static) analysis tools can also benefit, though off-chain tools can already deal with existing code, so the impact is smaller.
+
+A non-exhaustive list of future improvements which could benefit from this format:
+- Including a `JUMPDEST`-table (to avoid analysis at execution time) or removing `JUMPDEST`s entirely.
+- Introducing static jumps (with relative addresses) and jump tables, and disallowing dynamic jumps at the same time.
+- Requiring code section(s) to be terminated by `STOP`. (Assumptions like this can provide significant speed improvements in interpreters, such as a speed up of ~7% seen in [evmone](https://github.com/ethereum/evmone/pull/295).)
+- Multi-byte opcodes without any workarounds.
+- Representing functions as individual code sections instead of subroutines.
+- Introducing a specific section for the [EIP-2938 Account Abstraction](./eip-2938.md) "top-level AA execution frame", simplifying the proposal.
+
+## Specification
+
+*We use [RFC2119](https://tools.ietf.org/html/rfc2119) keywords in this section.*
+
+In order to guarantee that every EOF-formatted contract in the state is valid, we need to prevent already deployed (and not validated) contracts from being recognized as such format. This will be achieved by choosing a byte sequence for the *magic* that doesn't exist in any of the already deployed contracts. To prevent the growth of the search space and to limit the analysis to the contracts existing before these changes, we split changes into two hard forks, and disallow the starting byte of the format (the first byte of the magic) in the first hard fork.
```

Comment:
This is motivational, and should be moved to that section of the EIP rather than the specification section.  That being said, this feels a bit redundant to what is already in the motivation section so consider just dropping it.

Reply:
Actually, this looks like it is saying that this EIP has a hard dependency on EIP-3541.  Recommend removing the backlink from 3541 to this EIP, and making this EIP depend on 3541.  Then you can just delete this whole paragraph.